generator/src/main/java/com/graphicsfuzz/generator/tool/CustomMutatorSever.java

```diff
@@ -0,0 +1,129 @@
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+
+/**
+ * Server that can accept shaders from a libFuzzer custom mutator and send back a mutated shader.
+ */
+public class CustomMutatorSever {
+  public static void main(String[] args) {
+    try {
+      // TODO(metzman): If we don't switch from TCP, allow this to be configurable.
+      runServer(8666);
+    } catch (IOException | ParseTimeoutException | InterruptedException exception) {
+      exception.printStackTrace();
+      System.exit(1);
+    }
+  }
+
+  private static void runServer(int port)
+      throws IOException, ParseTimeoutException, InterruptedException {
+    byte[] headerBuff = new byte[28];
+    ServerSocket serverSocket = new ServerSocket(port);
+    Socket socket = serverSocket.accept();
+    InputStream inputStream = socket.getInputStream();
+    OutputStream outputStream = socket.getOutputStream();
+    while (true) {
+      // TODO(metzman): Figure out a better way to handle waiting for the header to arrive.
+      while (inputStream.available() < headerBuff.length) {
+        ;
+      }
+      inputStream.read(headerBuff, 0, headerBuff.length);
+      ByteBuffer headerByteBuffer = ByteBuffer.wrap(headerBuff);
+      headerByteBuffer.order(ByteOrder.LITTLE_ENDIAN);
+
+      long size1 = headerByteBuffer.getLong();
+      // Support two sizes for now so that one message format can be used for both CustomMutate and
+      // CustomCrossOver.
+      long size2 = headerByteBuffer.getLong();
+      assert size2 == 0;
+
+      // Java won't allow us to create an array with a "long" size. Therefore, we must convert it to
+      // int before creating the array. This is probably a non-issue because libFuzzer is unlikely
+      // to ever give us a shader larger than Integer.MAX_VALUE (e.g. ~2G).
+      assert size1 <= Integer.MAX_VALUE && size2 <= Integer.MAX_VALUE;
+
+      // GraphicsFuzz can't do anything with this unfortunately.
+      long maxOutSize = headerByteBuffer.getLong();
+
+      int libfuzzerSeed = headerByteBuffer.getInt();
+
+      // Read the Shader.
+      byte[] inputShaderBuff = new byte[((int) size1) + ((int) size2)];
+      socket.getInputStream().read(inputShaderBuff, 0, inputShaderBuff.length);
+      String inputShader = new String(inputShaderBuff);
+
+      try {
+        // Now mutate the shader.
+        TranslationUnit tu = ParseHelper.parse(inputShader, ShaderKind.FRAGMENT);
+        Mutate.mutate(tu, new RandomWrapper(libfuzzerSeed));
+        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+        try (PrintStream stream = new PrintStream(byteArrayOutputStream, true, "UTF-8")) {
+          PrettyPrinterVisitor.emitShader(
+              tu,
+              Optional.empty(),
+              stream,
+              PrettyPrinterVisitor.DEFAULT_INDENTATION_WIDTH,
```

Comment:
It occurs to me that if minimzing shader size matters, an indentation width of zero might be a good thing.

Reply:
I'm not sure I understand but I'm not sure eliminating whitespace is a helpful minimization here.
I think minimal testcases are most useful because the odds that random, totally arbitrary mutations result in new behavior are better when the testcase is reduced. But in structure aware fuzzing this isn't always important since the fuzzer knows not to mutate whitespace. What do you think?

This comment also makes me realize that single testcase minimization won't work well since the best we can do is hope mutations reduce the size of shaders, but this issue can be fixed later.

Reply:
I just meant that if the overhead of transferring shaders between libfuzzer and graphicsfuzz is high, then minimizing size of the shader text could help a bit, and minimizing whitespace might help with that.  A shader with deeply-nested conditionals can end up having a lot of whitespace in its text representation due to indentation.  But perhaps this is a non-issue in practice.

Reply:
OK Done